File: Lab3OOP/UserInterface.cpp
#include "UserInterface.h"
#include <iostream>
#include <iomanip>
#include <vector>
#include "Song.h"
void UserInterface::displayMessage(const std::string& message) {
    std::cout << message << std::endl;
}

void UserInterface::displayMenu() {
    std::cout << "\n1. Add composition\n";
    std::cout << "2. Show favorite compositions\n";
    std::cout << "3. Remove composition\n";
    std::cout << "4. Show all compositions\n";
    std::cout << "5. Exit\n";
    std::cout << "Select an option: ";
}

int UserInterface::displayCompositionListAndGetChoice(const std::vector<Composition*>& compositions) {
    int choice;
tryAgain:
    for (size_t i = 0; i < compositions.size(); ++i) {
        std::cout << std::setw(2) << i + 1 << ". "
            << compositions[i]->getTitle()
            << " (" << compositions[i]->getDuration() << " sec)\n";
    }
    std::cout << "Enter the index of the composition to remove (0 to cancel): ";
    while (true) {
        std::cin >> choice;

        
        if (std::cin.fail()) {
            std::cin.clear(); 
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); 
            std::cerr << "Error: Invalid input. Please enter a valid number." << std::endl << std::endl;
            goto tryAgain;
        }
        else {
            break;  
        }
    }
    return choice;
}

int UserInterface::getChoice() {
    int choice;
    std::cin >> choice;
    if (std::cin.fail()) {
        std::cin.clear();  
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); 
        displayError("Invalid input. Please enter a valid number.");
        return -1;  
    }
    return choice;
}
void UserInterface::displayCompositionList(const std::vector<Composition*>& compositions, const std::set<size_t>& chosen) {
    for (size_t i = 0; i < compositions.size(); ++i) {
        std::cout << std::setw(2) << i + 1 << ". ";
        if (chosen.count(i)) {
            std::cout << "* "; 
        }
        std::cout << compositions[i]->getTitle()
            << " (" << compositions[i]->getDuration() << " sec)\n";
    }
}

void UserInterface::displayFavoritesSummary(const std::set<size_t>& chosen, const std::vector<Composition*>& compositions) {
    int totalDuration = 0;
    for (auto index : chosen) {
        totalDuration += compositions[index]->getDuration();
    }
    std::cout << "\nTotal listening time for selected compositions: " << totalDuration << " seconds.\n";

    for (auto index : chosen) {
        if (const Song* song = dynamic_cast<const Song*>(compositions[index])) {
            std::cout << "Vocalist: " << song->getVocalist() << "\n";
        }
    }
    std::cout << std::endl;
}
void UserInterface::getCompositionDetails(std::string& title, int& duration, std::string& detail, bool isSong) {
    std::cout << "Enter name: ";
    std::cin >> title;


    while (true) {
        std::cout << "Enter duration (in seconds): ";
        std::cin >> duration;  
        if (std::cin.fail()) {
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');  
            std::cout << "Invalid input. Please enter a number for the duration." << std::endl;
        }
        else {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            break;
        }
    }

    if (isSong) {
        std::cout << "Enter the name of the vocalist: ";
    }
    else {
        std::cout << "Enter the name of the composer: ";
    }
    std::getline(std::cin, detail);  
}

void UserInterface::displayError(const std::string& errorMessage)  const {
    std::cerr << "Error: " << errorMessage << std::endl;
}
void UserInterface::exceptionMessage(const std::exception& e) {
    std::cerr << e.what() << std::endl;  
}

File: Lab3_oop_python/user_interface.py
from instrumental_work import InstrumentalWork
from song import Song
import sys

class UserInterface:
    def display_message(self, message: str):
        print(message)

    def display_menu(self):
        print("\n1. Add composition\n2. Show favorite compositions\n3. Remove composition\n4. Show all compositions\n5. Exit\nSelect an option: ")

    def get_choice(self):
        try:
            choice = int(input())
            return choice
        except ValueError:
            print("Error: Invalid input. Please enter a number.", file=sys.stderr)
            return None
        
    def display_composition_list_and_get_choice(self, compositions):
        try: 
            for i, comp in enumerate(compositions, start=1):
                print(f"{i}. {comp.get_title()} ({comp.get_duration()} sec)")
            index = int(input("Enter the index of the composition to remove (0 to cancel): "))
            return index
        except ValueError:
            print("Error: Invalid input. Please enter a valid number.", file=sys.stderr)
            return None


    def display_composition_list(self, compositions, chosen):
        for i, comp in enumerate(compositions, start=1):
            marker = "*" if i - 1 in chosen else " "
            print(f"{marker} {i}. {comp.get_title()} ({comp.get_duration()} sec)")

    def display_favorites_summary(self, chosen, compositions):
        total_duration = sum(compositions[index].get_duration() for index in chosen)
        print(f"\nTotal listening time for selected compositions: {total_duration} seconds.\n")

        for index in chosen:
            comp = compositions[index]
            if isinstance(comp, Song):
                print("Vocalist:", comp.get_vocalist())

    def get_composition_details(self, is_song):
        title = input("Enter name: ")

        while True:
            try:
                duration = int(input("Enter duration (in seconds): "))
                break
            except ValueError:
                print("Error: Duration must be a number. Please try again.", file=sys.stderr)

        if is_song:
            print("Enter the name of the vocalist: ")
        else:
            print("Enter the name of the composer: ")

        detail = input()
        return title, duration, detail
    
    def exceptionMessage(self, e):
        print(e, file=sys.stderr)
        
    def valueErrorMessage(self):
        print("Error: Wrong choice, try again", file=sys.stderr)

